chainerrl/agents/dqn.py

```diff
@@ -385,50 +382,80 @@ def act(self, obs):
+            assert self.last_action is not None
+            # Add a transition to the replay buffer
+            transition = {
+                'state': self.last_state,
+                'action': self.last_action,
+                'reward': reward,
+                'next_state': obs,
+                'is_state_terminal': False,
+            }
+            if self.recurrent:
+                transition['recurrent_state'] =\
+                    self.model.get_recurrent_state_at(
+                        self.train_prev_recurrent_states,
+                        0, unwrap_variable=True)
+                self.train_prev_recurrent_states = None
+                transition['next_recurrent_state'] =\
+                    self.model.get_recurrent_state_at(
+                        self.train_recurrent_states, 0, unwrap_variable=True)
+            self.replay_buffer.append(**transition)
+
+        # Update the target network
+        if self.t % self.target_update_interval == 0:
+            self.sync_target_network()
+
+        # Update the model
+        self.replay_updater.update_if_necessary(self.t)
+
+        # Choose an action
         with chainer.using_config('train', False), chainer.no_backprop_mode():
-            action_value = self.model(
-                self.batch_states([obs], self.xp, self.phi))
+            action_value =\
+                self._evaluate_model_and_update_train_recurrent_states([obs])
             q = float(action_value.max.array)
             greedy_action = cuda.to_cpu(action_value.greedy_actions.array)[0]
+        action = self.explorer.select_action(
+            self.t, lambda: greedy_action, action_value=action_value)
 
         # Update stats
         self.average_q *= self.average_q_decay
         self.average_q += (1 - self.average_q_decay) * q
 
-        self.logger.debug('t:%s q:%s action_value:%s', self.t, q, action_value)
-
-        action = self.explorer.select_action(
-            self.t, lambda: greedy_action, action_value=action_value)
         self.t += 1
-
-        # Update the target network
-        if self.t % self.target_update_interval == 0:
-            self.sync_target_network()
-
-        if self.last_state is not None:
-            assert self.last_action is not None
-            # Add a transition to the replay buffer
-            self.replay_buffer.append(
-                state=self.last_state,
-                action=self.last_action,
-                reward=reward,
-                next_state=obs,
-                next_action=action,
-                is_state_terminal=False)
-
         self.last_state = obs
         self.last_action = action
 
-        self.replay_updater.update_if_necessary(self.t)
-
+        self.logger.debug('t:%s q:%s action_value:%s', self.t, q, action_value)
         self.logger.debug('t:%s r:%s a:%s', self.t, reward, action)
 
         return self.last_action
 
+    def _evaluate_model_and_update_train_recurrent_states(self, batch_obs):
```

Comment:
Can we merge `_evaluate_model_and_update_train_recurrent_states` and `_evaluate_model_and_update_test_recurrent_states` and add a mandatory `test` parameter?